src/operations/execute_operation.ts

```diff
@@ -188,111 +164,113 @@ function executeWithServerSelection<TResult>(
     selector = readPreference;
   }
 
-  const serverSelectionOptions = { session };
-  function retryOperation(originalError: MongoError) {
-    const isWriteOperation = operation.hasAspect(Aspect.WRITE_OPERATION);
-    const isReadOperation = operation.hasAspect(Aspect.READ_OPERATION);
+  const server = await topology.selectServerAsync(selector, { session });
 
-    if (isWriteOperation && originalError.code === MMAPv1_RETRY_WRITES_ERROR_CODE) {
-      return callback(
-        new MongoServerError({
-          message: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
-          errmsg: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
-          originalError
-        })
-      );
+  if (session == null || !operation.hasAspect(Aspect.RETRYABLE)) {
+    // No session or non-retryable operation, simple early exit
+    try {
+      return await operation.executeAsync(server, session);
```

Comment:
optional for clarity
```suggestion
      return await operation.executeAsync(server, null);
```

Reply:
The session can still exist here so this would break non-retryable operations